wasm/instructions.h

```diff
@@ -25,27 +29,75 @@ struct BlockType {
     [[nodiscard]] bool operator==(BlockType const &) const = default;
 };
 
+constexpr std::string to_string(BlockType const &bt) {
+    std::string out;
+
+    if (ValueType const *v = std::get_if<ValueType>(&bt.value)) {
+        out += "(result ";
+        out += wasm::to_string(*v);
+        out += ")";
+    } else if (TypeIdx const *t = std::get_if<TypeIdx>(&bt.value)) {
+        out += "(type ";
+        out += std::to_string(*t);
+        out += ")";
+    } else {
+        assert(std::holds_alternative<BlockType::Empty>(bt.value));
+    }
+
+    return out;
+}
+
 struct MemArg {
     static std::optional<MemArg> parse(std::istream &);
 
     std::uint32_t align{};
     std::uint32_t offset{};
+
     [[nodiscard]] bool operator==(MemArg const &) const = default;
 };
 
+constexpr std::string to_string(MemArg const &ma, std::optional<std::uint32_t> natural_alignment = std::nullopt) {
+    std::string out;
+
+    // If offset == 0, omit offset
+    if (ma.offset != 0) {
+        out += "offset=";
+        out += std::to_string(ma.offset);
+    }
+
+    // Natural alignment, omit "align=" phrase
+    if (natural_alignment.has_value() && natural_alignment == ma.align) {
```

Comment:
Empty optionals always compare false with values, so you don't need to check has_value before comparing.